api/envoy/config/listener/v3/listener.proto

```diff
@@ -275,4 +280,20 @@ message Listener {
   // :ref:`use_original_dst <envoy_api_field_config.listener.v3.Listener.use_original_dst>`
   // to true. Default is true.
   google.protobuf.BoolValue bind_to_port = 26;
+
+  // Used to represent an internal listener which does not listen on OSI L4 address but can be used by the
+  // :ref:`envoy cluster <envoy_v3_api_msg_config.cluster.v3.Cluster>` to create a user space connection to.
+  // The internal listener acts as a tcp listener. It supports listener filters and network filter chains.
+  // The internal listener require :ref:`address <envoy_api_field_config.listener.v3.Listener.address>` has
+  // field `envoy_internal_address`.
+  //
+  // There are some limitations are derived from the implementation. The known limitations include
+  //
+  // * :ref:`ConnectionBalanceConfig <envoy_api_msg_config.listener.v3.Listener.ConnectionBalanceConfig>` is not
+  //   allowed because both cluster connection and listener connection must be owned by the same dispatcher.
+  // * :ref:`tcp_backlog_size <envoy_api_field_config.listener.v3.Listener.tcp_backlog_size>`
+  // * :ref:`freebind <envoy_api_field_config.listener.v3.Listener.freebind>`
+  // * :ref:`transparent <envoy_api_field_config.listener.v3.Listener.transparent>`
+  // [#not-implemented-hide:]
+  InternalListenerConfig internal_listener = 27;
```

Comment:
Per my comment in the other ticket, can we put this inside a oneof "listener_type" or similar, and then we can later work on fixing this proto to move the other listener types inside this oneof via deprecations?

Reply:
added. Even though we want to choose exactly one from "tcp, udp, api, internal", but we cannot set `required = true` annotation because of compatibility. Let me know if there is an advanced technology to resolve it.